storage/src/merkle_storage.rs

```diff
@@ -0,0 +1,937 @@
+        tree.iter().for_each(|(k, v)| {
+            hasher.update(&self.encode_irmin_node_kind(&v.node_kind)).expect("hasher");
+            hasher.update(&[k.len() as u8]).expect("hasher");
+            hasher.update(&k.clone().into_bytes()).expect("hasher");
+            hasher.update(&(HASH_LEN as u64).to_be_bytes()).expect("hasher");
+            hasher.update(&v.entry_hash).expect("hasher");
+        });
+
+        hasher.finalize().unwrap().as_ref().try_into().expect("EntryHash conversion error")
+    }
+
+    fn hash_blob(&self, blob: &ContextValue) -> EntryHash {
+        let mut hasher = State::new(HASH_LEN, None).unwrap();
+        hasher.update(&(blob.len() as u64).to_be_bytes()).expect("Failed to update hasher state");
+        hasher.update(blob).expect("Failed to update hasher state");
+
+        hasher.finalize().unwrap().as_ref().try_into().expect("EntryHash conversion error")
+    }
+
+    fn encode_irmin_node_kind(&self, kind: &NodeKind) -> Vec<u8> {
+        match kind {
+            NodeKind::NonLeaf => vec![0, 0, 0, 0, 0, 0, 0, 0],
+            NodeKind::Leaf => vec![255, 0, 0, 0, 0, 0, 0, 0],
+        }
+    }
+
+
+    fn get_tree(&self, hash: &EntryHash) -> Result<Tree, MerkleError> {
+        match self.get_entry(hash)? {
+            Entry::Tree(tree) => Ok(tree),
+            Entry::Blob(_) => Err(MerkleError::FoundUnexpectedStructure {
+                sought: "tree".to_string(),
+                found: "blob".to_string(),
+            }),
+            Entry::Commit { .. } => Err(MerkleError::FoundUnexpectedStructure {
+                sought: "tree".to_string(),
+                found: "commit".to_string(),
+            }),
+        }
+    }
+
+    fn get_commit(&self, hash: &EntryHash) -> Result<Commit, MerkleError> {
+        match self.get_entry(hash)? {
+            Entry::Commit(commit) => Ok(commit),
+            Entry::Tree(_) => Err(MerkleError::FoundUnexpectedStructure {
+                sought: "commit".to_string(),
+                found: "tree".to_string(),
+            }),
+            Entry::Blob(_) => Err(MerkleError::FoundUnexpectedStructure {
+                sought: "commit".to_string(),
+                found: "blob".to_string(),
+            }),
+        }
+    }
+
+    fn get_entry(&self, hash: &EntryHash) -> Result<Entry, MerkleError> {
+        match self.staged.get(hash) {
+            None => {
+                let entry_bytes = self.db.get(hash)?;
+                match entry_bytes {
+                    None => Err(MerkleError::EntryNotFound { hash: HashType::ContextHash.bytes_to_string(hash) } ),
+                    Some(entry_bytes) => Ok(bincode::deserialize(&entry_bytes)?),
+                }
+            }
+            Some(entry) => Ok(entry.clone()),
+        }
+    }
+
+    fn get_non_leaf(&self, hash: EntryHash) -> Node {
+        Node { node_kind: NodeKind::NonLeaf, entry_hash: hash }
+    }
+
+    fn get_db<P: AsRef<Path>>(path: P, cache: &Cache) -> DB {
```

Comment:
@michal-stanek 
can you please check this?
http://ci.tezedge.com/simplestaking/tezedge/1597/1/3
![image](https://user-images.githubusercontent.com/8159517/97054425-0eea4580-1585-11eb-9744-76a311da3fa6.png)